Amazon.QLDB.Driver/driver/QldbDriver.cs

```diff
@@ -211,12 +265,165 @@ public IEnumerable<string> ListTableNames()
         /// </summary>
         public void Dispose()
         {
-            this.sessionPool.Dispose();
+            if (!this.isClosed)
+            {
+                this.isClosed = true;
+                while (this.sessionPool.Count > 0)
+                {
+                    this.sessionPool.Take().Close();
+                }
+
+                this.sessionPool.Dispose();
+            }
+        }
+
+        private T Execute<T>(Func<TransactionExecutor, T> func, RetryPolicy retryPolicy, Action<int> retryAction)
+        {
+            if (this.isClosed)
+            {
+                this.logger.LogError(ExceptionMessages.DriverClosed);
+                throw new QldbDriverException(ExceptionMessages.DriverClosed);
+            }
+
+            bool replaceDeadSession = false;
+            int retryAttempt = 0;
+            while (true)
+            {
+                QldbSession session = null;
+                try
+                {
+                    if (replaceDeadSession)
+                    {
+                        session = this.StartNewSession();
+                    }
+                    else
+                    {
+                        session = this.GetSession();
+                    }
+
+                    T returnedValue = session.Execute(func);
+                    this.ReleaseSession(session);
+                    return returnedValue;
+                }
+                catch (RetriableException re)
+                {
+                    // If initial session is invalid, always retry once with a new session.
+                    if (re.InnerException is InvalidSessionException && retryAttempt == 0)
```

Comment:
This logic seems almost entirely duplicated from the async case, can we refactor it to reduce duplication at all?

Reply:
Not in a way I believe that is helpful. Trying to function block out catch blocks doesn't work very well in this case.

Reply:
What about having the sync call to the async and waiting on it?